src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking_Fields.cs

```diff
@@ -0,0 +1,608 @@
+                // (1,7): warning CS8618: Non-nullable field 'F1' is uninitialized.
+                // class C
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "C").WithArguments("field", "F1").WithLocation(1, 7));
+        }
+
+        [Fact]
+        public void ReadWriteFields_ExplicitConstructors()
+        {
+            var source =
+@"class C
+{
+    internal object F1;
+    private object? F2;
+    private object?[] F3;
+    internal object[]? F4;
+    internal C()
+    {
+    }
+    internal C(object o)
+    {
+        F1 = o;
+        F3 = new[] { o, null };
+    }
+    internal C(object x, object y)
+    {
+        F2 = x;
+        F4 = new[] { x, y };
+    }
+}";
+            var comp = CreateStandardCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (7,14): warning CS8618: Non-nullable field 'F3' is uninitialized.
+                //     internal C()
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "C").WithArguments("field", "F3").WithLocation(7, 14),
+                // (7,14): warning CS8618: Non-nullable field 'F1' is uninitialized.
+                //     internal C()
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "C").WithArguments("field", "F1").WithLocation(7, 14),
+                // (15,14): warning CS8618: Non-nullable field 'F3' is uninitialized.
+                //     internal C(object x, object y)
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "C").WithArguments("field", "F3").WithLocation(15, 14),
+                // (15,14): warning CS8618: Non-nullable field 'F1' is uninitialized.
+                //     internal C(object x, object y)
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "C").WithArguments("field", "F1").WithLocation(15, 14));
+        }
+
+        [Fact]
+        public void ReadOnlyFields_DefaultConstructor()
+        {
+            var source =
+@"class C
+{
+#pragma warning disable 0169
+#pragma warning disable 0649
+    private readonly object F1;
+    internal readonly object? F2;
+    internal readonly object?[] F3;
+    private readonly object[]? F4;
+}";
+            var comp = CreateStandardCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (1,7): warning CS8618: Non-nullable field 'F3' is uninitialized.
+                // class C
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "C").WithArguments("field", "F3").WithLocation(1, 7),
+                // (1,7): warning CS8618: Non-nullable field 'F1' is uninitialized.
+                // class C
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "C").WithArguments("field", "F1").WithLocation(1, 7));
+        }
+
+        [Fact]
+        public void ReadOnlyFields_ExplicitConstructors()
```

Comment:
Could you add `internal readonly (object a, object) F5;` too?

Reply:
Added `Tuples` test.